### PR TITLE
Refactor DNS Cache for horizontal scaling

### PR DESCRIPTION
**Pull request by vkumra-broad**
- addresses [issue #563](https://github.com/DataBiosphere/leonardo/issues/563)
- see [Leo Horizontal Scaling Implementation Plan](https://docs.google.com/document/d/1KQ4BSEQvrvdArDVvr8dupO9t9g3ksql41FCONNh3Yeg) for context

### CODE REVIEW

**Comment by vkumra-broad** [comment]
jenkins retest

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dns/ClusterDnsCache.scala
Comment (kyuksel) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dns/ClusterDnsCache.scala]: For readability, I'd suggest refactoring the cache key into a key case class similarly to in `SamAuthProvider.notebookAuthCache`:

```scala
private case class DnsCacheKey(googleProject: GoogleProject, clusterName: ClusterName)

new CacheLoader[DnsCacheKey, Future[GetClusterResponse]] {
```

This would have an added benefit of making code with direct tuple indices more readable; e.g. `getActiveClusterByName(key.googleProject, key.clusterName)` instead of `getActiveClusterByName(key._1, key._2)`.

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dns/ClusterDnsCache.scala
Comment (kyuksel) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dns/ClusterDnsCache.scala]: Update the comment above since we don't have a `GetByProjectAndName` message anymore

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dns/ClusterDnsCache.scala
Comment (kyuksel) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dns/ClusterDnsCache.scala]: This naming throws me off. I wonder if it'd be more intuitive to rename `GetClusterResponse` as `NotebookHostStatus` or `HostStatus`, and its descendants `NotebookHostNotFound` or `HostNotFound`, etc?

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dns/ClusterDnsCache.scala
Comment (kyuksel) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dns/ClusterDnsCache.scala]: I wonder if we can do away with `HostToIp` and store its contents directly in the Guava cache. We could define something like

```scala
case class DnsCacheValue(host: Host, ip: IP)
```

and then update `CacheLoader.load()` accordingly?

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dns/ClusterDnsCache.scala
Comment (kyuksel) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dns/ClusterDnsCache.scala]: If this is still relevant after refactoring based on the other comments, we may want to rewrite as something like

```scala
projectNameToHostEntry(cluster) match { 
  case (_, ClusterReady(host)) => ??? 
}
```

**Review by kyuksel — CHANGES_REQUESTED** [review]

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dns/ClusterDnsCache.scala
Comment (kyuksel) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dns/ClusterDnsCache.scala]: We should probably make ^ configurable, too.

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dns/ClusterDnsCache.scala
Comment (vkumra-broad) [reply src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dns/ClusterDnsCache.scala]: Agreed, I came up with that number considering  that when a notebook is opened most of the call would happen within that time period or less. It would also prevent the cache from having stale data. I was planning on using load test results to tweak that duration.

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dns/ClusterDnsCache.scala
Comment (vkumra-broad) [reply src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dns/ClusterDnsCache.scala]: I am not sure if this would work, the JupyterNameService talks directly to the ClusterDnsCache object.

**Comment by vkumra-broad** [comment]
jenkins retest

**Comment by vkumra-broad** [comment]
jenkins retest

**Comment by vkumra-broad** [comment]
jenkins retest

**Comment by vkumra-broad** [comment]
jenkins retest

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dns/ClusterDnsCache.scala
Comment (kyuksel) [reply src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dns/ClusterDnsCache.scala]: Ah right. @rtitle, do you know if we have a way to get around that seeming limitation where `JupyterNameService` talks to `ClusterDnsCache` object implicitly via the descriptor. It would be great to not have to have that `var`.

File: src/test/resources/reference.conf
Comment (kyuksel) [inline src/test/resources/reference.conf]: Why did the port number change?

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dns/ClusterDnsCache.scala
Comment (kyuksel) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dns/ClusterDnsCache.scala]: We can use the newly created `DnsCacheKey` instead of `(GoogleProject, ClusterName)` so

```scala
  private def projectNameToHostEntry(c: Cluster): (DnsCacheKey, HostStatus) = ???
```

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dns/ClusterDnsCache.scala
Comment (kyuksel) [reply src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dns/ClusterDnsCache.scala]: I'd still recommend using pattern matching like ^^.

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dns/ClusterDnsCache.scala
Comment (kyuksel) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dns/ClusterDnsCache.scala]: Is the above `import` used?

**Review by kyuksel — CHANGES_REQUESTED** [review]
Some more feedback. Getting close :) I'd like to still look at the test changes.

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dns/ClusterDnsCache.scala
Comment (kyuksel) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dns/ClusterDnsCache.scala]: Already talked to Vikram about this but for a reminder for myself...we're gonna want to add the new entry rather than overwriting the Map content.

Comment (kyuksel) [reply src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dns/ClusterDnsCache.scala]: I also wonder if we can write a unit test that would fail with the code line above and pass with the change?

File: src/test/scala/org/broadinstitute/dsde/workbench/leonardo/dns/ClusterDnsCacheSpec.scala
Comment (kyuksel) [inline src/test/scala/org/broadinstitute/dsde/workbench/leonardo/dns/ClusterDnsCacheSpec.scala]: `.copy()` above looks redundant since it's not overwriting any of the `Cluster` fields.

File: src/test/scala/org/broadinstitute/dsde/workbench/leonardo/dns/ClusterDnsCacheSpec.scala
Comment (kyuksel) [inline src/test/scala/org/broadinstitute/dsde/workbench/leonardo/dns/ClusterDnsCacheSpec.scala]: `GcsPathUtils` unused

File: src/test/scala/org/broadinstitute/dsde/workbench/leonardo/dns/ClusterDnsCacheSpec.scala
Comment (kyuksel) [inline src/test/scala/org/broadinstitute/dsde/workbench/leonardo/dns/ClusterDnsCacheSpec.scala]: `Matchers` unused?

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dns/ClusterDnsCache.scala
Comment (kyuksel) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dns/ClusterDnsCache.scala]: I wonder if `projectClusterToHost` would be more self-documenting.

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dns/ClusterDnsCache.scala
Comment (kyuksel) [reply src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dns/ClusterDnsCache.scala]: Actually, `projectNameToHostEntry` is called from only within `projectNameToHost.load()` where only the second element (`._2`) of the output tuple is used. Therefore, I'd suggest changing the method as:

```scala
private def clusterToHostStatus(c: Cluster): HostStatus = {
  if (c.status.isStartable)
    HostPaused
  else if (c.dataprocInfo.hostIp.isDefined)
    HostReady(host(c))
  else
    HostNotReady
}
```

File: src/test/scala/org/broadinstitute/dsde/workbench/leonardo/dns/ClusterDnsCacheSpec.scala
Comment (kyuksel) [inline src/test/scala/org/broadinstitute/dsde/workbench/leonardo/dns/ClusterDnsCacheSpec.scala]: Let's make sure the references to `ProjectNameToHost` and `HostToIp` are accurate at the end, in terms of naming and upper/lower case initials.

File: src/test/scala/org/broadinstitute/dsde/workbench/leonardo/dns/ClusterDnsCacheSpec.scala
Comment (kyuksel) [inline src/test/scala/org/broadinstitute/dsde/workbench/leonardo/dns/ClusterDnsCacheSpec.scala]: Can we add another test with a cluster that `isStartable` and verify that its `HostStatus` ends up being `HostPaused` and its `ClusterDnsCache` entry is `None`?

File: src/test/scala/org/broadinstitute/dsde/workbench/leonardo/dns/ClusterDnsCacheSpec.scala
Comment (kyuksel) [inline src/test/scala/org/broadinstitute/dsde/workbench/leonardo/dns/ClusterDnsCacheSpec.scala]: Is it possible to set the Guava cache's `cacheExpiryTime` low enough to also test that the cache is flushed after that much time as expected?

**Review by kyuksel — CHANGES_REQUESTED** [review]
I think I'm through now with the review.

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/HttpJupyterDAO.scala
Comment (kyuksel) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/HttpJupyterDAO.scala]: Getting the compilation warning
```
[warn] /Users/kyuksel/github/leonardo/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/HttpJupyterDAO.scala:35:21: method apply in trait LoadingCache is deprecated: see corresponding Javadoc for more information.
[warn]     clusterDnsCache.projectNameToHost(DnsCacheKey(googleProject, clusterName)).mapTo[HostStatus]
[warn]                     ^
```
which can be fixed by using `get` instead of `apply`:
```scala
clusterDnsCache.projectNameToHost.get(DnsCacheKey(googleProject, clusterName)).mapTo[HostStatus]
```

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/ProxyService.scala
Comment (kyuksel) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/ProxyService.scala]: Getting the compilation warning
```
[warn] /Users/kyuksel/github/leonardo/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/ProxyService.scala:255:21: method apply in trait LoadingCache is deprecated: see corresponding Javadoc for more information.
[warn]     clusterDnsCache.projectNameToHost(DnsCacheKey(googleProject, clusterName)).mapTo[HostStatus]
[warn]                     ^
```
which can be fixed by using `get` instead of `apply` (which is called implicitly with the current code):
```scala
clusterDnsCache.projectNameToHost.get(DnsCacheKey(googleProject, clusterName)).mapTo[HostStatus]
```

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dns/ClusterDnsCache.scala
Comment (kyuksel) [reply src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dns/ClusterDnsCache.scala]: I've made [this PR](https://github.com/DataBiosphere/leonardo/pull/627) to this PR's branch to encapsulate the `volatile` and `synchronized` stuff so the client code is cleaner. We could also utilize it for other similar situations in the codebase.

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/ClusterMonitorActor.scala
Comment (kyuksel) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/ClusterMonitorActor.scala]: Note to self to double-check that removing ^ doesn't introduce a regression.

**Comment by kyuksel** [comment]
Closing until it's ready for re-review/testing

**Review by kyuksel — COMMENTED** [review]
Changing my reviewer status to commenter

**Review by kyuksel — APPROVED** [review]

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/ClusterMonitorActor.scala
Comment (kyuksel) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/ClusterMonitorActor.scala]: Removed the method above and renamed `JupyterProxyDAO.getStatus` as `isProxyAvailable` and started calling that here directly because

1. with the call `ensureClusterReadyForProxying(ip, clusterStatus)` no longer necessary, this method would've been down to the `jupyterProxyDAO.getStatus` call
2. `JupyterProxyDAO.getStatus` was not returning us a `status` per se but whether the proxy is available or not

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/HttpJupyterDAO.scala
Comment (kyuksel) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/HttpJupyterDAO.scala]: Removed the `isProxyAvailable` method that used to be in `ClusterMonitorActor` and renamed `JupyterProxyDAO.getStatus` as `isProxyAvailable` and started calling that here directly because

1. with the call `ensureClusterReadyForProxying(ip, clusterStatus)` no longer necessary, this method would've been down to the `jupyterProxyDAO.getStatus` call
2. `JupyterProxyDAO.getStatus` was not returning us a `status` per se but whether the proxy is available or not

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/HttpJupyterDAO.scala
Comment (kyuksel) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/HttpJupyterDAO.scala]: Trying to be more consistent with the usage of `Host` in the Jupyter proxying context

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/ClusterMonitorActor.scala
Comment (kyuksel) [reply src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/ClusterMonitorActor.scala]: We're no longer trying to cache IPs as soon as hosts become available upon cluster creation so `ensureClusterReadyForProxying` (whose name is a bit misleading) is no longer necessary.

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/ValueBox.scala
Comment (kyuksel) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/ValueBox.scala]: ^^ is a project I worked on where `ValueBox` was useful so I wanted to spread the love ❤️

File: src/test/scala/org/broadinstitute/dsde/workbench/leonardo/dns/ClusterDnsCacheSpec.scala
Comment (kyuksel) [inline src/test/scala/org/broadinstitute/dsde/workbench/leonardo/dns/ClusterDnsCacheSpec.scala]: Renamed the test clusters to make their purposes more explicit and set the above's status to `Creating` from the default `Unknown`

File: src/test/scala/org/broadinstitute/dsde/workbench/leonardo/dns/ClusterDnsCacheSpec.scala
Comment (kyuksel) [inline src/test/scala/org/broadinstitute/dsde/workbench/leonardo/dns/ClusterDnsCacheSpec.scala]: Using `CacheStats` (enabled at the time of cache creation via `recordStats` in `ClusterDnsCache.scala`) to make sure the expected number of cache loadings, misses and evictions happen throughout the test. I was motivated to do this particularly because of the `500` errors we saw recently that looked like a DNS cache. 

From what I read, enabling stats recording shouldn't cause a noticeable performance penalty.

**Comment by kyuksel** [comment]
Comparing some manual performance testing on my branch FIAB to the prod version using the ASHG notebook, I have
- not noticed a difference between cell executions
- notebook loading takes ~3x longer (10.5s vs 3.5s) with the new changes, culprit call being 
```
/notebooks/{project}/{cluster}/api/contents/kyuksel-dnscache-testing/Large_Notebook_File_Example.ipynb?type=notebook&_=1541784790843
```

I'm in the process of creating a second FIAB with the dev version of Leo, to eliminate potential env/network-related confounders for the performance degradation.

**Comment by kyuksel** [comment]
Following up on the previous note:

The performance comparison between the two FIABs (the one with the branch vs. the one with dev) did not yield a statistically significant difference (I was nerdy enough to run a t-test, which resulted in a p-value of .33 🤓) and the branch actually had a 1s lower mean. Therefore I feel comfortable enough to merge this PR (once we're done reviewing it).

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dns/ClusterDnsCache.scala
Comment (kyuksel) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dns/ClusterDnsCache.scala]: Made `projectClusterToHostStatus` `private` and `def`ined getter for it so the cache isn't writable outside the class.

Comment (rtitle) [reply src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dns/ClusterDnsCache.scala]: But isn't it still be writable via the `cache` method? External callers could do `.cache.put(...)`.

Normally we've not exposed the cache directly at all, but defined a getter like:
```
def getHostStatus(googleProject: GoogleProject, clusterName: ClusterName): Future[HostStatus] = {
  projectClusterToHostStatus.get(DnsCacheKey(googleProject, clusterName)
}
```

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/ClusterMonitorActor.scala
Comment (rtitle) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/ClusterMonitorActor.scala]: Why do we need to update the ip in the database here? I think `clusterQuery.setToRunning` already does that.

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/ValueBox.scala
Comment (rtitle) [reply src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/ValueBox.scala]: Cool - maybe it could go in workbench-libs?

**Review by rtitle — APPROVED** [review]
Looks good, just a couple comments above. Good to know that performance didn't degrade with the small cache expiry time. I'm interested whether this resolves the issue we've been seeing with the current cache, where `Stopped` cluster status does not get reflected and leads to 500 errors to the caller.

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/ValueBox.scala
Comment (kyuksel) [reply src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/ValueBox.scala]: I agree. Will try to factor it out in a separate PR.

**Comment by kyuksel** [comment]
@rtitle Thanks for the great feedback! I think I've addressed all but the workbench-libs one, which I'll target as a separate PR. Rerunning the integration tests. Did you want to take another look before I merge?

**Review by rtitle — APPROVED** [review]
Latest changes look good! :shipit: if tests pass

**Comment by kyuksel** [comment]
jenkins retest

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/ClusterMonitorActor.scala
Comment (kyuksel) [reply src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/ClusterMonitorActor.scala]: It turns out, we need to update the database because later on we rely on the database to get the host information to determine `ifProxyAvailable`, or else, we think that the host isn't available and get stuck checking the status. This happens when we resume clusters and the cluster transitions from `Starting` to `Running`.

**Pull request merged**
